uint256/uint256.go

```diff
@@ -0,0 +1,142 @@
+	}
+	if value.Cmp(minUint256) == -1 {
+		return nil, fmt.Errorf("cannot set Uint256 to %v as it underflows min value of %v", value, minUint256)
+	}
+	u.value = *new(big.Int).Set(&value)
+	return u, nil
+}
+
+// Copy sets the underlying value of u to a copy of the given Uint256 param, and returns the modified receiver struct
+func (u *Uint256) Copy(v *Uint256) *Uint256 {
+	u.value = *new(big.Int).Set(&v.value)
+	return u
+}
+
+// Cmp calls the underlying Cmp method for the big.Int stored in a Uint256 struct as its value field
+func (u *Uint256) Cmp(v *Uint256) int {
+	return u.value.Cmp(&v.value)
+}
+
+// Equals returns true if the two Uint256 structs have the same underlying values, false otherwise
+func (u *Uint256) Equals(v *Uint256) bool {
+	return u.Cmp(v) == 0
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Add(augend, addend *Uint256) (*Uint256, error) {
+	sum := new(big.Int).Add(&augend.value, &addend.value)
+	return u.Set(*sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Sub(minuend, subtrahend *Uint256) (*Uint256, error) {
+	difference := new(big.Int).Sub(&minuend.value, &subtrahend.value)
+	return u.Set(*difference)
+}
+
+// Mul sets u to multiplicand * multiplier and returns u as the product
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Mul(multiplicand, multiplier *Uint256) (*Uint256, error) {
+	product := new(big.Int).Mul(&multiplicand.value, &multiplier.value)
+	return u.Set(*product)
+}
+
+// String returns the string representation for Uint256 structs
+func (u *Uint256) String() string {
+	return u.value.String()
+}
+
+// MarshalJSON implements the json.Marshaler interface
+// it specifies how to marshal a Uint256 struct so that it can be written to disk
+func (u Uint256) MarshalJSON() ([]byte, error) {
+	return []byte(u.value.String()), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+// it specifies how to unmarshal a Uint256 struct so that it can be reconstructed from disk
+func (u *Uint256) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		return nil
+	}
+
+	var value big.Int
+	_, ok := value.SetString(string(b), 10)
+	if !ok {
+		return fmt.Errorf("not a valid integer value: %s", b)
+	}
+	u.value = value
+	return nil
+}
+
+// EncodeRLP implements the rlp.Encoder interface
+// it makes sure the `value` field is encoded even though it is private
+func (u *Uint256) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, &u.value)
+}
+
+// DecodeRLP implements the rlp.Decoder interface
+// it makes sure the `value` field is decoded even though it is private
+func (u *Uint256) DecodeRLP(s *rlp.Stream) error {
+	return s.Decode(&u.value)
```

Comment:
Check range after decoding as well?

Reply:
done.

may i ask why it's not necessary to implement a range check in `EncodeRLP` or `MarshalJSON`? is it overkill?

Reply:
I think it's overkill because the processing is when we decode - **unless** we send to other peers with different implementations, but then it's not our responsibility. Bullet proof would be to do it here as well, but I don't think it's really needed. Give it a thought and decide yourself.

Reply:
i've given it a thought, and decided to leave it in its current state. 

outputting an error when marshaling or encoding will not solve the problem of having an invalid variable in the first place—which is not true in the case of unmarshaling and decoding.